doc/neps/nep-0038-SIMD-optimizations.rst

```diff
@@ -180,20 +195,50 @@ Anyone writing a SIMD loop will use the ``npyv_load_u32`` macro instead of the
 architecture specific intrinsic. The code also supplies guard macros for
 compilation and runtime, so that the proper loops can be chosen.
 
+At compile time, the CPU is probed to determine which features are available.
+These are mapped into the CPU-architecture-specific dictionary
+``__cpu_features__``, the values are filled in at run-time by again probing
+the CPU.
+
+Two new build options are available to ``runtests.py`` and ``setup.py build``.
+The absolute minimum required features to compile are defined by
+``--cpu-baseline``.  For instance, on ``x86_64`` this defaults to ``SSE3``. If
+this minimum is not met, build will fail. The
```

Comment:
the build should never fail under any circumstances except one thing if the user provided unknown CPU features.
the minimum features will be enabled if the compiler support it.

Reply:
the minimum features are set to depend on the wide range of CPUs support, 
in other words,  almost all `x86`(32-bit) CPUs supports `SSE SSE2` and all `x86` CPUs that run on 64-bit mode supports `SSE SSE2 SSE3`. and here is the full list from source:
```Python
conf_min_features = dict(
        x86 = "SSE SSE2",
        x64 = "SSE SSE2 SSE3",
        ppc64 = '', # play it safe
        ppc64le = "VSX VSX2",
        armhf = '', # play it safe
        aarch64 = "NEON NEON_FP16 NEON_VFPV4 ASIMD"
)
````

Reply:
If they are running on a CPU that does not support `SSE3`, and `SSE3` is specified as the minimum, what will happen?

Reply:
in #13516, there's a check for this case raising a Python runtime error, but let's assume this validation doesn't exist or somehow the compiler generates instructions related to this feature before the validation. well, in this case, a crash must happen with "illegal instruction error".
however, some compilers like msvc and intel already enable these features by default, also GCC do the same behavior with aarch64(arm8) and ppc64le.

Reply:
So the correct phrasing would be "If this minimum is not recognized by the compiler, build will fail".

I guess runtime will always succeed, there are always the non-SIMD fallbacks. This means there will be either one three alternatives: we will always provide the non-SIMD loops, and perhaps provide exactly two additional ones: a minimum and a maximum. At some point in the future we could decide to drop the non-SIMD ones.

Reply:
> I guess runtime will always succeed, there are always the non-SIMD fallbacks

Here yes, but we have places where we use SSE2 (IIRC not SSE3) without any runtime dispatch; it's a hard minimum.

Reply:
> So the correct phrasing would be "If this minimum is not recognized by the compiler, build will fail".

If these minimum features aren't recognized by the compiler, these features will be skipped during distutils testing process, so the correct phrasing would be " the minimum features will be enabled if the compiler supports it."

I think we need to clarify how the whole progress work through writing a decent documentation, I have to continue work on [#43](https://github.com/mattip/numpy/pull/43) as soon as possible.

the weird thing is almost all of the discussions are only focused on universal intrinsics and ignored the infrastructure. from my way of thinking the great challenge was the infrastructure.
the universal intrinsics are not even close to considered an issue since our rule is just guiding the compilers and then let them do their optimization job. that's why I think universal intrinsics will allow any developer to easily write SIMD code.

Reply:
> the weird thing is almost all of the discussions are only focused on universal intrinsics and ignored the infrastructure. from my way of thinking the great challenge was the infrastructure.

Yes that's a good observation. I think you may be correct that the infrastructure is perhaps the harder part. Reasons for the asymmetry in discussion may be that Raghuveer is spending a lot of time on intrinsics and wants to make sure he isn't blocked, and that build/dispatching/release infrastructure is generally a topic that people tend to shy away from.



> I think we need to clarify how the whole progress work through writing a decent document, I have to continue work on [#43](https://github.com/mattip/numpy/pull/43) as soon as possible.

Thanks for the ping on that, I'll have a look now.

Reply:
So I think this NEP is about ready, and then https://github.com/mattip/numpy/pull/43 can be moved into the reference docs. We're then in a good place to start testing of the main PR I'd think.

Reply:
adopted "the minimum features will be enabled if the compiler support it". I will begin a PR to add mattip/numpy#43 to the documentation.

Reply:
@seiko2plus, I apologize for causing the bias in the discussion. I really had no feedback to give for runtime and compile time infrastructure because most of it already looked good to me. My expertise is mainly programming with AVX and that's pretty much what I can talk about :)

Reply:
@r-devulap no need to apologize, your input is much appreciated!

Reply:
@r-devulap, I'm not an expert too actually I'm not even a python coder lol. we're just discovering something new, and together we gonna make it works. com'n I feel like we're a family here and as @rgommers said no need to apologize, your inputs are too valuable and appreciated.